### PR TITLE
MAINT: Fix ReDOS vulnerability in crackfortran.py

### PR DESCRIPTION
**Pull request by molsonkiko**
Address ReDOS vulnerability flagged by CodeQL first mentioned in #23338.

There was an capture group in `crackfortran.nameargspattern` that used an unchecked `.*` when it could have instead used `[^@\s]*` to achieve the same result.

### CODE REVIEW

File: numpy/f2py/crackfortran.py [hunk 1]
```diff
@@ -935,7 +935,7 @@
     r'(?:,(?P<attributes>[\w(),]+))?(::)?(?P<name>\b[a-z$_][\w$]*\b)'
     r'(?:\((?P<params>[\w,]*)\))?\Z', re.I)
 nameargspattern = re.compile(
-    r'\s*(?P<name>\b[\w$]+\b)\s*(@\(@\s*(?P<args>[\w\s,]*)\s*@\)@|)\s*((result(\s*@\(@\s*(?P<result>\b[\w$]+\b)\s*@\)@|))|(bind\s*@\(@\s*(?P<bind>.*)\s*@\)@))*\s*\Z', re.I)
+    r'\s*(?P<name>\b[\w$]+\b)\s*(@\(@\s*(?P<args>[\w\s,]*)\s*@\)@|)\s*((result(\s*@\(@\s*(?P<result>\b[\w$]+\b)\s*@\)@|))|(bind\s*@\(@\s*(?P<bind>[^\s@]*)\s*@\)@))*\s*\Z', re.I)
```
[the remark this answers is not in the document]
Comment (molsonkiko) [reply numpy/f2py/crackfortran.py:938]: disregard this check. Here's what I found using my local installation.

```py
from numpy.f2py.crackfortran import nameargspattern
for ii in range(12, 20):
    %timeit nameargspattern.search('@)@bind@(@' *ii)
```
results:
```
31.4 µs ± 4.07 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
36.4 µs ± 4.76 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
37.5 µs ± 2.76 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
43.5 µs ± 4.49 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
52.4 µs ± 8.39 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
53.8 µs ± 4.62 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
63 µs ± 7.03 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
64.8 µs ± 3.68 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

**Comment by charris** [comment]
Got flagged again.

**Comment by molsonkiko** [comment]
Yeah, it's still far from perfect, but it takes *a lot* more repititions of `@)@bind@(@` to get terrible performance.

Empirically it seems like each instance of `@)@bind@(@` is now increasing runtime by about 10%, as opposed to the approximate doubling I was previously seeing.

**Comment by molsonkiko** [comment]
nice way to do diagnostics:
```py
'''
script for testing numpy.f2py.crackfortran.nameargspattern on adversarially
designed repeated "@)@bind@(@" input.
Example usage:
> python time_crackfortran.py 20 40 1
'''
import time
from numpy.f2py.crackfortran import nameargspattern
import matplotlib.pyplot as plt

bad_thing = '@)@bind@(@'

def time_n_bad_thing_reps(n_reps: int, timeout: float) -> float:
    total_time = 0
    n_trials = 0
    while total_time < timeout:
        s = bad_thing * n_reps
        t0 = time.perf_counter()
        nameargspattern.search(s)
        time_this_trial = time.perf_counter() - t0
        total_time += time_this_trial
        n_trials += 1
        if total_time > timeout:
            break
    return total_time/n_trials

def plot_bad_thing_times(min_n_reps, max_n_reps, timeout):
    reps_nums = list(range(min_n_reps, max_n_reps))
    times = [time_n_bad_thing_reps(n_reps, timeout)*1000
        for n_reps in reps_nums]
    plt.scatter(reps_nums, times)
    plt.xlabel('number of reps of "@)@bind@(@"')
    plt.ylabel('avg time (ms) to for re.search to execute')
    plt.title('''performance testing for crackfortran nameargspattern
on adversarially designed input''')
    plt.tight_layout()
    plt.show()
    
if __name__ == '__main__':
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('min_n_reps', type=int, default=20, nargs='?')
    parser.add_argument('max_n_reps', type=int, default=40, nargs='?')
    parser.add_argument('timeout', type=float, default=1., nargs='?')
    args = parser.parse_args()
    plot_bad_thing_times(args.min_n_reps, args.max_n_reps, args.timeout)
```
and here's what I get on my machine:
time_crackfortran

That's time in *milliseconds*. I'd say the problem is fixed.

**Comment by charris** [comment]
Nice improvement. I dismissed the original alert and we can dismiss the current one if needed.

**Comment by charris** [comment]
One of the benchmark failures is unrelated, but the other is the new test. I suspect testing  time taken is tough to get right.

**Comment by molsonkiko** [comment]
So I thought of a reason why the regex in my previous commit isn't good: it rejects all whitespace characters, even though maybe some of those should have been accepted.

**Comment by molsonkiko** [comment]
Agree that testing time taken is hard. I think that if the sample size is large enough, the variance of average runtime goes down to the point where we don't need to worry about super weird problems. That's why I increased the number of trials per rep count from 32 to 128.

Since I'm not the final arbiter of what counts as "acceptable" performance from f2py, I'm happy with however you want to rewrite the tests.

**Comment by charris** [comment]
> I'm happy with however you want to rewrite the tests.

I'm actually more worried about not breaking the parsing :) That is a hard to understand bit of code and it is likely the tests aren't as complete as they should be.

**Comment by charris** [comment]
close/reopen

**Comment by charris** [comment]
Let's give this a shot. If you think of a better solution, feel free to make another PR.

**Comment by molsonkiko** [comment]
Awesome! I don't really think I'm qualified to fool around with f2py, beyond my optimization of that regex.

I can take a look at other regexes in crackfortran to see if there are other optimizations that could be made, but I doubt I'll find any.

I feel like my main takeaway from this experience is that the automated CodeQL warnings could probably be improved without too much trouble, like identifying a possible replacement for the problematic region of a regex, rather than just indicating a possible failure mode.

**Comment by charris** [comment]
> CodeQL warnings could probably be improved without too much trouble

That would be nice. There were several other regex alerts that gave more useful information.

File: numpy/f2py/tests/test_crackfortran.py [hunk 2]
```diff
@@ -276,3 +277,49 @@ class TestUnicodeComment(util.F2PyTest):
...
+                # the problematic pattern.
+                times_median_doubled += median > 2 * last_median
+                # also try to rule out non-exponential but still bad cases
+                # arbitrarily, we should set a hard limit of 10ms as too slow
+                assert median < trials_per_count * 0.01
+            last_median = median
+        # we accept that maybe the median might double once, due to
+        # the CPU scheduler acting weird or whatever. More than that
+        # seems suspicious.
+        assert times_median_doubled < 2
```
Comment (seberg) [inline numpy/f2py/tests/test_crackfortran.py:325]: Saw this test fail at least once recently on CI.  We may want to make it more robust, or maybe also just remove it, at least if that happens more often.

**Pull request merged**